Mathlib/Algebra/GroupWithZero/NonZeroDivisors.lean

```diff
@@ -277,3 +275,89 @@ lemma nonZeroSMulDivisors_mulOpposite_eq_op_nonZeroDivisors :
     Rᵐᵒᵖ ⁰[R] = R⁰.op := rfl
 
 end nonZeroSMulDivisors
+
+noncomputable section Associated
+
+open nonZeroDivisors
+
+section MonoidWithZero
+
+variable (α : Type*) [MonoidWithZero α]
+
+/-- The units of the monoid of non-zero divisors of `α` are equivalent to the units of `α`. -/
+def nonZeroDivisorsUnitsEquiv : α⁰ˣ ≃* αˣ :=
+  MulEquiv.ofBijective (Units.map α⁰.subtype) ⟨Units.map_injective Subtype.val_injective,
+    fun u ↦ ⟨IsUnit.unit ⟨⟨⟨u, IsUnit.mem_nonZeroDivisors u.isUnit⟩, ⟨(u⁻¹ : αˣ),
+        IsUnit.mem_nonZeroDivisors u⁻¹.isUnit⟩, by simp, by simp⟩, rfl⟩,
+      by rw [Units.ext_iff, IsUnit.unit_of_val_units, Units.coe_map, Submonoid.coe_subtype]⟩⟩
+
+@[simp]
+theorem nonZeroDivisorsUnitsEquiv_apply (u : α⁰ˣ) :
+    nonZeroDivisorsUnitsEquiv α u = (u : α) := rfl
+
+@[simp]
+theorem nonZeroDivisorsUnitsEquiv_symm_apply (u : αˣ) :
+    ((nonZeroDivisorsUnitsEquiv α).symm u : α) = (u : α) := by
+  obtain ⟨v, rfl⟩ := (nonZeroDivisorsUnitsEquiv α).surjective u
+  rw [MulEquiv.symm_apply_apply, nonZeroDivisorsUnitsEquiv_apply]
+
+end MonoidWithZero
+
+section CommMonoidWithZero
+
+variable (α : Type*) [CommMonoidWithZero α]
+
+theorem associatesMk_mem_nonZeroDivisors_iff (a : α) :
+    Associates.mk a ∈ (Associates α)⁰ ↔ a ∈ α⁰ := by
+  rw [mem_nonZeroDivisors_iff, mem_nonZeroDivisors_iff, ← not_iff_not]
+  push_neg
+  constructor
+  · rintro ⟨⟨x⟩, hx₁, hx₂⟩
+    refine ⟨x, ?_, ?_⟩
+    · rwa [← Associates.mk_eq_zero, ← Associates.mk_mul_mk, ← Associates.quot_mk_eq_mk]
+    · rwa [← Associates.mk_ne_zero, ← Associates.quot_mk_eq_mk]
+  · refine fun ⟨b, hb₁, hb₂⟩ ↦ ⟨Associates.mk b, ?_, by rwa [Associates.mk_ne_zero]⟩
+    rw [Associates.mk_mul_mk, hb₁, Associates.mk_zero]
+
+/-- To any class in `Associates α⁰`, one can associate a class in `(Associates α)⁰` by sending a
+representative `a : α⁰` of the class to the class of `(a : α)`. The map obtained is in fact a
+`MulEquiv`, see `AssociatesNonZeroDivisorsMulEquiv`. -/
+def AssociatesNonZeroDivisorsMonoidHom :
+    Associates α⁰ →* (Associates α)⁰ where
+  toFun := Quotient.lift (fun ⟨x, _⟩ ↦  ⟨Associates.mk x, by
+      rwa [associatesMk_mem_nonZeroDivisors_iff]⟩) (by
+    rintro _ _ ⟨v, hv⟩
+    rw [Subtype.mk.injEq, Associates.mk_eq_mk_iff_associated]
+    exact ⟨nonZeroDivisorsUnitsEquiv α v, by
+      rw [nonZeroDivisorsUnitsEquiv_apply, ← hv, Submonoid.coe_mul]⟩)
+  map_one' := rfl
+  map_mul' x y := Quotient.inductionOn₂ x y fun _ _ ↦ rfl
+
+@[simp]
+theorem AssociatesNonZeroDivisorsMonoidHom_apply (a : α⁰) :
+    (AssociatesNonZeroDivisorsMonoidHom α ⟦a⟧ : Associates α) = Associates.mk (a : α) := rfl
+
+/-- This is the `MulEquiv` version of `AssociatesNonZeroDivisorsMonoidHom`. -/
+def AssociatesNonZeroDivisorsMulEquiv :
+    Associates α⁰ ≃* (Associates α)⁰ := by
```

Comment:
Similarly, this can be produced fully computably.

Reply:
Tell me if you need help with this one

Reply:
Yeah, that one is a bit trickier. I'll try to make it work and I'll get back to you if I can't. Thanks.

Reply:
(And feel free to push your changes too)

Reply:
Done. Tell me if you like it.

Reply:
Well, it is a lot cleaner. Thanks!